provider/k8s/k8s_discover.go

```diff
@@ -0,0 +1,214 @@
+// Package k8s provides pod discovery for Kubernetes.
+package k8s
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"path/filepath"
+	"strconv"
+
+	"github.com/ericchiang/k8s"
+	corev1 "github.com/ericchiang/k8s/apis/core/v1"
+	"github.com/ghodss/yaml"
+
+	"github.com/hashicorp/go-multierror"
+	"github.com/mitchellh/go-homedir"
+)
+
+const (
+	// AnnotationKeyPort is the annotation name of the field that specifies
+	// the port name or number to append to the address.
+	AnnotationKeyPort = "hashicorp.com/consul-auto-join-port"
+)
+
+type Provider struct{}
+
+func (p *Provider) Help() string {
+	return `Kubernetes (K8S):
+
+    provider:         "k8s"
+    kubeconfig:       Path to the kubeconfig file.
+    namespace:        Namespace to search for pods (defaults to "default").
+    label_selector:   Label selector value to filter pods.
+    field_selector:   Field selector value to filter pods.
+    host_network:     "true" if pod host IP and ports should be used.
+
+    The kubeconfig file value will be searched in the following locations:
+
+     1. Use path from "kubeconfig" option if provided.
+     2. Use path from KUBECONFIG environment variable.
+     3. Use default path of $HOME/.kube/config
+
+    By default, the Pod IP is used to join. The "host_network" option may
+    be set to use the Host IP. No port is used by default. Pods may set
```

Comment:
May be use "HostIP" instead of "Host IP" to keep things consistent.

Reply:
I use "Pod IP" above so I think this is consistent. I assume folks aren't too familiar with the K8S API so didn't want to use the direct field names.